c/src/proactor/epoll.c

```diff
@@ -1426,10 +1419,10 @@ static void pconnection_tick(pconnection_t *pc) {
   pn_transport_t *t = pc->driver.transport;
   if (pn_transport_get_idle_timeout(t) || pn_transport_get_remote_idle_timeout(t)) {
     ptimer_set(&pc->timer, 0);
-    uint64_t now = pn_i_now2();
-    uint64_t next = pn_transport_tick(t, now);
+    int64_t now = pn_proactor_now_64();
+    int64_t next = pn_transport_tick(t, now);
     if (next) {
-      ptimer_set(&pc->timer, next - now);
+      ptimer_set(&pc->timer, (uint64_t) next - now);
     }
```

Comment:
Yes, ok. But first I'll spend some time trying to remember why I was so eager on signed arithmetic back then.

Reply:
Signed arithmetic is a lot easier in general, since you can have negative intermediate results in an expression, and express offsets in both directions in a straightforward way. I'd prefer signed in the public API  unless there's a massively good reason to go unsigned. Overflow in 34,000,000,000 years instead of 17,000,000,000 years  doesn't strike me as a massively good reason, although obviously programmers in the year 17,000,001,969 will curse my short-sightedness. Trust Andrew to be considerate of the hyper-intelligent horseshoe-crabs that will be masters of the planet then.

Reply:
@jdanekrh If it helps the unsigned arithmetic is still correct when the cast signed values are negative. And given that the C language only defines unsigned arithmetic to be correctly wrapping the subtraction must use uints.

@alanconway you are assuming that the now value is based at 0 which is not a given - and in fact it is common to test timer wraparound by starting the 'now' monotonic timer at MAXINT-a few minutes.

Reply:
To be clear I agree that the external API should maintain signed values - it really has to in any case for back compatibility. but the very specific case of this subtraction needs to be done in correct wraparound mode so unsigned is required here (even if you just cast it back immediately to signed)

Reply:
@astitcher  and I are in violent agreement, just like old times. Signed in the API, and the @astitcher unsigned magic in the impl is correct. Speculation about horseshoe crabs out-evolving humans in the distant future is irrelevant (they probably already have)

Reply:
> The more I think about this, the more confused I feel.

I feel your pain.